---
title: "unknown-species-id-template"
output:
  pdf_document: default
  html_document: default
---

June 2023
Anita Wray
Code originally written by Diana B

Walkthrough of species id for unknown samples:

Using the rds file output from microhaplot:
1. read in rds files
2. apply read depth filters
3. apply allele balance filter
```{r setup}
knitr::opts_knit$set(root.dir = "~/Desktop/RE_BS/stock review 2025 work/")
```

```{r warning=FALSE, include=FALSE}
library(tidyverse)
library(readxl)
library(stringr)
library(lubridate)
library(rubias)
library(ggpattern)
library(forcats)
```

```{r warning=FALSE}

source("R/rockfish-funcs2.R")

# get the names of the files
fdf <- read.table("rds-file-list.txt", stringsAsFactors = FALSE, header = TRUE) %>%
  tibble::as_tibble()

dir <- "microhaplot/"



# cycle over them, read them and add the gtseq_run column on each.
# at the end, bind them together.
genos_long <- lapply(1:nrow(fdf), function(i) {
  message("Working on ", fdf$file[i])
  call_genos_from_haplotRDS(path = file.path(dir, fdf$file[i])) %>%
    mutate(gtseq_run = fdf$gtseq_run[i]) %>%
    select(gtseq_run, everything())
}) %>%
  bind_rows()

#genos_long$id <- gsub('-','',genos_long$id)

# we go ahead and save it in data/processed, with xz compression
saveRDS(genos_long, file = "processed/called_genos_anita.rds", compress = "xz")


#### In the end, let us get a data frame that includes genotypes for all the individuals  ####
# and which explicitly has NAs in places where data are missing, and also 
# has the NMFS_DNA_ID on there
genos_long_explicit_NAs <- genos_long %>%
  select(gtseq_run, id) %>%
  unique() %>%
  unite(col = gid, sep = "_", gtseq_run, id) %>%
  select(gid) %>%
  unlist() %>%
  unname() %>%
  expand.grid(gid = ., locus = unique(genos_long$locus), gene_copy = 1:2, stringsAsFactors = FALSE) %>%
  tibble::as_tibble() %>% 
  separate(gid, into = c("gtseq_run", "id"), convert = TRUE, sep = '_') %>%
  left_join(., genos_long) %>%
  arrange(gtseq_run, id, locus, gene_copy)

# and then save that
saveRDS(genos_long_explicit_NAs, file = "processed/called_genos_na_explicit_anita.rds", compress = "xz")

final_df <- data.frame(matrix(NA, nrow = 90, ncol = 2))
colnames(final_df) <- c("locus_name", 'count of missing alleles')

unique_locus_names <- unique(genos_long$locus)
j <- 1
for (i in unique_locus_names){
  temp_df <- subset(genos_long, genos_long$locus == i)
  final_df[j,1] <- i
  final_df[j,2] <- sum(is.na(temp_df$allele))
  j <- j +1
}

ggplot(data = final_df, aes(x = fct_reorder(locus_name, `count of missing alleles`), y = `count of missing alleles`/2)) +
  geom_col() +
  ylab('Number of samples missing allele')+
  xlab('Locus')+
  theme_classic()
  
```


Using those genotypes...
```{r}
genos_long_explicit_NAs %>%
  group_by(gtseq_run, id) %>%
  tally() 
```


Look at missing data:
180 gene copies total (90 loci x2)
```{r}
ind_to_toss <- genos_long_explicit_NAs %>%
  group_by(gtseq_run, id) %>%
  filter(is.na(allele)) %>% # missing data
  tally() %>%
  arrange(desc(n)) %>% # remove samples with >30% missing data
  filter(n > 54)

##write those removed samples for the PCA
write.csv(ind_to_toss, file = "~/Desktop/RE_BS/stock review 2025 work/removed_samples_rubias_04_11_25.csv")

# remove those from the df
genos_ind_filtered <- genos_long_explicit_NAs %>%
  anti_join(., ind_to_toss)

table(length(unique(genos_ind_filtered$id)))
  
```



Load baseline data
```{r baseline-data}
# baseline data - curated, 997 indivs
baseline <- readRDS("new_baseline_data/processed/sebastes_spp_id_baseline_haplotypes_04_17.rds")

# remove the 6 loci that had HWE and other issues
to_remove <- read_csv("data/loci_to_remove.csv")

baseline90 <- baseline %>%
 anti_join(., to_remove)

# remind myself which species are in the baseline:
baseline90 %>%
  select(collection) %>%
  unique() %>%
  arrange()

tossers <- baseline90 %>%
  select(indiv, gtseq_run, id) %>%
  unique() %>%
  group_by(indiv) %>%
  tally() %>%
  filter(n >1)

baseline90_one_each <- baseline90 %>%
  anti_join(., tossers)

# baseline data - curated, 1024 indivs
baseline_spp_info <- baseline90_one_each %>%
  select(sample_type, repunit, collection, indiv, gtseq_run, id, species) %>% 
  unique()
baseline_spp_info$gtseq_run <- as.character(baseline_spp_info$gtseq_run)
```



```{r baseline-format}
# slim that down to just the matching field with the unknowns
for_alleidx <- baseline90_one_each %>%
 select(-indiv, -c(1:3, 12:13), -species)

for_alleidx$gtseq_run <- as.character(for_alleidx$gtseq_run)
```



```{r}
genos_ind_filtered <- genos_ind_filtered %>%
  subset(!id %in% for_alleidx$id)


# merge the two dataframes
merged_df <- bind_rows(for_alleidx, genos_ind_filtered)

# first make integers of the alleles
alle_idxs <- merged_df %>% 
  dplyr::select(gtseq_run, id, locus, gene_copy, allele) %>%
  group_by(locus) %>%
  mutate(alleidx = as.integer(factor(allele, levels = unique(allele)))) %>%
  ungroup() %>%
  arrange(gtseq_run, id, locus, alleidx) # rubias can handle NA's, so no need to change them to 0's

  
# and spread the alleles
two_col <- alle_idxs %>%
  group_by(id, locus) %>%
  unite(loc, locus, gene_copy, sep = ".") %>%
  ungroup() %>%
  select(-allele) %>%
  pivot_wider(names_from = loc, values_from = alleidx) 


```



add back on info for reference and make two-column format for rubias
```{r}
# baseline
reference <- two_col %>%
  left_join(., baseline_spp_info) %>%
  filter(!is.na(species)) %>%
  select(-gtseq_run, -id, -species) %>%
  select(sample_type, repunit, collection, indiv, everything()) 

```

```{r}
# mixture
rubias_mix <- two_col %>%
  anti_join(., baseline_spp_info) %>%
  mutate(sample_type = "mixture", collection = "krista", repunit = NA) %>%
  select(sample_type, repunit, collection, everything()) %>%
  unite(gtseq_run, id, col = "indiv", sep = "_") #removed gtseq_run, id

#Pull out two known sunset individual
croc_known <- c('gtseq3_H-14-MI-V0270','gtseq3_H-14-MI-V0272')

croc <- subset(rubias_mix, rubias_mix$indiv %in% croc_known)
croc$sample_type <- 'reference'
croc$repunit <- 'crocotulus'
croc$collection <- 'crocotulus'

#Add it to the reference 
reference <- rbind(reference, croc)

#Remove it from the mixture
rubias_mix <- subset(rubias_mix, !rubias_mix$indiv %in% croc_known)

```

## Mixture assignment with rubias


```{r run-rubias}
rubias_output <- infer_mixture(reference = reference, mixture = rubias_mix, gen_start_col = 5)

```



```{r}
# take the top output for each sample
top_assign <- rubias_output$indiv_posteriors %>%
  group_by(indiv) %>%
  slice_max(., order_by = PofZ)

nonREBS <- top_assign %>%
  subset(repunit != 'melanostictus') %>%
  subset(repunit != 'aleutianus')

table(nonREBS$repunit)

df <- apply(top_assign,2,as.character)
write.csv(df, file = '~/Desktop/RE_BS/stock review 2025 work/rubias_output_04_17_2025.csv')
```

Check on z-scores:

```{r echo=FALSE}
top_assign %>%
  ggplot(aes(x = z_score)) +
  geom_histogram()

ggsave("Zscores_04_17_2025.pdf")
```

```{r eval=FALSE, include=FALSE}
df <- top_assign %>% 
  select(c('indiv','repunit','collection','z_score', 'n_miss_loci')) 

df$indiv <-  sub("gtseq3_", "", df$indiv)

spp_indiv <- read.csv('species_ID.csv') %>%
  select(c('Sample_Name', 'Genetic_ID'))
colnames(spp_indiv) <- c('indiv','species')

df_compare <- merge(df, spp_indiv, by = 'indiv')
```